### PR TITLE
Annotations with same predicate are ordered before applying primary t…

### PR DESCRIPTION
**Pull request by Eng3lFT**
# Description

## What

Orders the annotations before setting the primary theme of the Index to be sure is always the same.

## Why

[JIRA ticket](https://financialtimes.atlassian.net/browse/UPPSF-4128).

## Scope and particulars of this PR (Please tick all that apply)

- [ ] Tech hygiene (dependency updating & other tech debt)
- [X] Bug fix
- [ ] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: pkg/mapper/mapper.go [hunk 1]
```diff
@@ -81,16 +83,68 @@ func (h *Handler) ToIndexModel(enrichedContent schema.EnrichedContent, contentTy
...
+		return
+	}
+	annotation := sorted[0]
+	annIDs := ids[annotation.ID]
+	conceptTypes := h.Config.ConceptTypes
+	for _, taxonomy := range annotation.Types {
+		switch taxonomy {
+		case conceptTypes.Get("organisation"):
+			if annotation.Predicate == predicate {
+				setPrimaryTheme(model, annotation.PrefLabel, getCmrIDWithFallback(tmeOrganisations, annIDs))
```
Comment (atanasdinov) [inline pkg/mapper/mapper.go:109]: The only thing changing is the first argument of `getCmrIDWithFallback`. I'd suggest to:
- have a single predicate check
- use the switch only to set the TME argument
- have a single `setPrimaryTheme` call.

File: pkg/mapper/mapper.go [hunk 2]
```diff
@@ -81,16 +83,68 @@ func (h *Handler) ToIndexModel(enrichedContent schema.EnrichedContent, contentTy
...
+func sortAnnotations(annotations []schema.Thing, ids map[string][]string, predicate string) []schema.Thing {
+	var store []schema.Thing
+
+	for _, annotation := range annotations {
+		if annotation.Predicate == predicate && ids[annotation.ID] != nil {
+			store = append(store, annotation)
+		}
+	}
+	sort.Slice(store, func(i, j int) bool {
+		return store[i].ID < store[j].ID
```
Comment (atanasdinov) [inline pkg/mapper/mapper.go:140]: Sorting by ID (i.e. by `"http://api.ft.com/things/{uuid}"` values) seems as arbitrary to me as not having any sorting at all. I'd pick `PrefLabel` or something else as the criteria.

File: pkg/mapper/mapper.go [hunk 3]
```diff
@@ -81,16 +83,68 @@ func (h *Handler) ToIndexModel(enrichedContent schema.EnrichedContent, contentTy
...
+			if annotation.Predicate == predicate {
+				setPrimaryTheme(model, annotation.PrefLabel, getCmrIDWithFallback(tmeRegions, annIDs))
+				return
+			}
+		}
+	}
+}
+
+func sortAnnotations(annotations []schema.Thing, ids map[string][]string, predicate string) []schema.Thing {
+	var store []schema.Thing
```
Comment (atanasdinov) [inline pkg/mapper/mapper.go:132]: A nitpick but the naming here is off. Could we improve it?

File: pkg/mapper/mapper.go [hunk 4]
```diff
@@ -81,16 +83,68 @@ func (h *Handler) ToIndexModel(enrichedContent schema.EnrichedContent, contentTy
...
+
+	for _, annotation := range annotations {
```
Comment (atanasdinov) [inline pkg/mapper/mapper.go:134]: A function called `sortAnnotations` shouldn't be the one responsible for such filtering. Could we rework this a little?

**Review by atanasdinov — CHANGES_REQUESTED** [review]

File: pkg/mapper/mapper.go (hunk 2 above)
Comment (Eng3lFT) [reply pkg/mapper/mapper.go:140]: Fair enough, TBH I picked `ID` just for ensure a sorting criteria.

File: pkg/mapper/mapper.go (hunk 4 above)
Comment (Eng3lFT) [reply pkg/mapper/mapper.go:130]: Totally agree, it does more than just sorting.

Comment (atanasdinov) [reply pkg/mapper/mapper.go:130]: Sorry if I wasn't clear enough. I thought it's better to split the filtering and sorting and not do it in the same function in the first place.

File: pkg/mapper/mapper.go [hunk 5]
```diff
@@ -81,16 +83,64 @@ func (h *Handler) ToIndexModel(enrichedContent schema.EnrichedContent, contentTy
...
+	sorted := filterSortAnnotations(annotations, ids, predicate)
+	if len(sorted) == 0 {
+		return
+	}
+	annotation := sorted[0]
+	annIDs := ids[annotation.ID]
+	conceptTypes := h.Config.ConceptTypes
+
+	for _, taxonomy := range annotation.Types {
+		if annotation.Predicate == predicate {
```
Comment (atanasdinov) [inline pkg/mapper/mapper.go:107]: Do we need this check if we've already filtered the annotations by this predicate?

File: test/testdata/exampleEnrichedContentModel.json [hunk 6]
```diff
@@ -96,19 +96,6 @@
         "apiUrl": "http://api.ft.com/things/b7ea3c33-ea8c-432e-bb7e-e3bbc8fdc2bb"
       }
     },
-    {
```
Comment (atanasdinov) [inline test/testdata/exampleEnrichedContentModel.json:99]: Do we have to drop this test data?

Comment (Eng3lFT) [reply test/testdata/exampleEnrichedContentModel.json:99]: The idea is to have an example without the predicate `http://www.ft.com/ontology/annotation/about` so `cmr_primarytheme` can be [empty](https://github.com/Financial-Times/content-rw-elasticsearch/pull/65/files/56a7d0572fe6b47ce4e8b900609c9e5d1529a42d#diff-bb8bd79eb951625575ebef7ad6b760931a3f35876ee4509d4c38fd0e4592df42R33).

File: pkg/mapper/mapper.go (hunk 4 above)
Comment (Eng3lFT) [reply pkg/mapper/mapper.go:130]: I would filter in a separate function and sorting may be done in the same function as are only three lines.

**Review by martin-stanchev — APPROVED** [review]

**Pull request merged**